verifiable-api/server/bin/handlers.rs

```diff
@@ -0,0 +1,439 @@
+    let filter: Filter = logs_filter_query.try_into().map_err(map_server_err)?;
+
+    // Fetch the filter logs from RPC
+    let logs = rpc.get_logs(&filter).await.map_err(map_server_err)?;
+
+    // Create receipt proofs for each log
+    let receipt_proofs = create_receipt_proofs_for_logs(&logs, rpc)
+        .await
+        .map_err(map_server_err)?;
+
+    Ok(Json(LogsResponse {
+        logs,
+        receipt_proofs,
+    }))
+}
+
+/// This method returns an array of all logs matching the filter with given id.
+/// Corresponding to each log, it also returns the transaction receipt and a Merkle proof of its inclusion.
+///
+/// Replaces the `eth_getFilterLogs` RPC method.
+pub async fn get_filter_logs<N: NetworkSpec, R: ExecutionRpc<N>>(
+    Path(filter_id): Path<U256>,
+    State(ApiState { rpc, .. }): State<ApiState<N, R>>,
+) -> Response<FilterLogsResponse<N>> {
+    // Fetch the filter logs from RPC
+    let logs = rpc
+        .get_filter_logs(filter_id)
+        .await
+        .map_err(map_server_err)?;
+
+    // Create receipt proofs for each log
+    let receipt_proofs = create_receipt_proofs_for_logs(&logs, rpc)
+        .await
+        .map_err(map_server_err)?;
+
+    Ok(Json(FilterLogsResponse {
+        logs,
+        receipt_proofs,
+    }))
+}
+
+/// This method returns the changes since the last poll for a given filter id.
+/// If filter is of logs type, then corresponding to each log,
+/// it also returns the transaction receipt and a Merkle proof of its inclusion..
+///
+/// Replaces the `eth_getFilterChanges` RPC method.
+pub async fn get_filter_changes<N: NetworkSpec, R: ExecutionRpc<N>>(
+    Path(filter_id): Path<U256>,
+    State(ApiState { rpc, .. }): State<ApiState<N, R>>,
+) -> Response<FilterChangesResponse<N>> {
+    let filter_changes = rpc
+        .get_filter_changes(filter_id)
+        .await
+        .map_err(map_server_err)?;
+
+    Ok(Json(match filter_changes {
+        FilterChanges::Logs(logs) => {
+            // Create receipt proofs for each log
+            let receipt_proofs = create_receipt_proofs_for_logs(&logs, rpc)
+                .await
+                .map_err(map_server_err)?;
+
+            FilterChangesResponse::Logs(FilterLogsResponse {
+                logs,
+                receipt_proofs,
+            })
+        }
+        FilterChanges::Hashes(hashes) => FilterChangesResponse::Hashes(hashes),
+        FilterChanges::Empty => FilterChangesResponse::Hashes(vec![]),
+        FilterChanges::Transactions(txs) => FilterChangesResponse::Hashes(
+            txs.into_iter().map(|t| t.inner.tx_hash().clone()).collect(),
+        ),
+    }))
+}
+
+/// This method returns a list of all accounts (along with their proofs) that are accessed by a given transaction.
+/// The list includes the `from`, `to` and `beneficiary` addresses as well.
+///
+/// Replaces the `eth_createAccessList` RPC method.
+pub async fn create_access_list<N: NetworkSpec, R: ExecutionRpc<N>>(
```

Comment:
One of the reasons we want this API is that eth_createAccessList doesn't work in some cases.

A better way to do this would be to simulate the call in revm and collect all accounts accessed instead of using eth_createAccessList.

Reply:
Yeah exactly. It's harder but `eth_createAccessList` is designed more for simulating a tx so that you can add an access list to your transaction. This means that in many client implementations it is stricter than `eth_call`. For example it checks that the from account has enough eth to pay for gas, which call usually doesn't do. This very often leads to this method failing to return a reasonable access list for our evm call to consume.